src/otaclient/create_standby/resume_ota.py

```diff
@@ -149,3 +190,142 @@ def resume_ota(self) -> None:
+                    file_size += read_size
+                    tmp_dst.write(bufferview[:read_size])
+                    hash_f.update(bufferview[:read_size])
+                os.posix_fadvise(src_fd, 0, 0, os.POSIX_FADV_DONTNEED)
+                os.posix_fadvise(tmp_dst_fd, 0, 0, os.POSIX_FADV_DONTNEED)
+
+            calculated_digest = hash_f.digest()
+            if calculated_digest != expected_digest:
+                self._remove_entry(tmp_dst_fname, dir_fd=dst_dir_fd)
+                return
+
+            try:
+                self._all_resource_digests.pop(calculated_digest)
+                # NOTE: both src and dst are under the same folder now
+                os.replace(
+                    tmp_dst_fname,
+                    expected_digest_hex,
+                    src_dir_fd=dst_dir_fd,
+                    dst_dir_fd=dst_dir_fd,
+                )
+                self._internal_que.put_nowait(file_size)
+
+            except KeyError:
+                # basically should not happen, as now we only scan resources presented in
+                #   the target OTA image now.
+                pass
+        finally:
+            self._se.release()
+            self._remove_entry(tmp_dst_fname, dir_fd=dst_dir_fd)
+
+    def resume_ota(self) -> None:
+        """Scan the OTA resource folder leftover by previous interrupted OTA."""
+        src_dir_fd = os.open(self._src_resource_dir, os.O_RDONLY)
+        dst_dir_fd = os.open(self._dst_resource_dir, os.O_RDONLY)
+
+        status_reporter_t = threading.Thread(
+            target=self._report_uploader_thread,
+            name="resume_ota_status_reporter",
+            daemon=True,
+        )
+        status_reporter_t.start()
+        # NOTE: create a shallow copy of all digests
+        _all_digests = set(self._all_resource_digests)
+        try:
+            with ThreadPoolExecutor(
+                max_workers=cfg.MAX_PROCESS_FILE_THREAD,
+                initializer=self._thread_initializer,
+            ) as pool:
+                _count = 0
+
+                for entry in os.scandir(self._src_resource_dir):
+                    _entry_digest_hex = entry.name
+                    if _entry_digest_hex == EMPTY_FILE_SHA256:
+                        continue
+
+                    try:
+                        _expected_digest = bytes.fromhex(_entry_digest_hex)
+                    except ValueError:
+                        continue
+
+                    if _expected_digest not in _all_digests:
+                        continue
+
+                    self._se.acquire()
+                    _count += 1
+                    pool.submit(
+                        self._process_resource_at_thread,
+                        _expected_digest,
+                        _entry_digest_hex,
+                        src_dir_fd=src_dir_fd,
+                        dst_dir_fd=dst_dir_fd,
+                    )
+            logger.info(f"totally {_count} of OTA resource files are scanned")
```

Comment:
Not comment for this PR. Do you think this value will be useful as metrics in future?

Reply:
The actual number of resources being scanned and used in fastpath can be recorded as metrics though, not the `_count` here. In the future I will record the number of actual accepted resources :+1: